Add warning about skipped vars

caleb15 opened the pull request:
### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Many people are surprised to find out that a variable is still registered even when skipped. See https://github.com/ansible/ansible/issues/33827 and https://github.com/ansible/ansible/issues/17500 and https://github.com/ansible/ansible/issues/4297.  It is a persistent problem that has tripped dozens of people up. This commit adds a warning so people are aware of the unintuitive behavior.
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Docs Pull Request

+label: docsite_pr


##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
Conditionals

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

caleb15 committed:
Add warning about skipped vars

##### SUMMARY
<!--- Your description here -->
Many people are surprised to find out that a variable is still registered even when skipped. See https://github.com/ansible/ansible/issues/33827 and https://github.com/ansible/ansible/issues/17500 and https://github.com/ansible/ansible/issues/4297.  It is a persistent problem that has tripped dozens of people up. This commit adds a warning so people are aware of the unintuitive behavior.

##### ISSUE TYPE
- Docs Pull Request

+label: docsite_pr

caleb15 commented:
You can also see people who have ran into this problem in google forums: https://groups.google.com/forum/#!msg/ansible-project/NGXcbyQvz-Y/lvc1Gw0fjX0J;context-place=forum/ansible-project

bcoca commented on docs/docsite/rst/user_guide/playbooks_conditionals.rst:
```diff
@@ -69,6 +69,7 @@ decide to do something conditionally based on success or failure::
 
 
 .. note:: both `success` and `succeeded` work (`fail`/`failed`, etc).
+.. warning:: Do NOT use `defined` or `default` to check if a variable was skipped. A skipped variable is still defined - metadata is inserted into the variable to record whether it was skipped, among other things.
```
the variable is not skipped, the task is, the register reflects the task status.

bcoca replied:
FYI this is already noted here `docs/docsite/rst/user_guide/playbooks_variables.rst`

caleb15 replied:
> the variable is not skipped, the task is, the register reflects the task status.

I changed the wording, does it look good now?

> FYI this is already noted here docs/docsite/rst/user_guide/playbooks_variables.rst

Good point. It was added in 4 years ago and people are still running into issues, so I think it deserves another mention.

bcoca replied:
i would change this to make a ref to that note,  not that i expect most to read either and still hit this issue.

caleb15 replied:
Done. Look good?

caleb15 replied:
@bcoca Not sure if you saw my earlier message - I added a ref to the note, is it good to merge now?

s-hertel replied:
@caleb15 The docs team will review and merge.

caleb15 committed:
change wording

caleb15 committed:
add reference to registering variables section

https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#registering-variables

samccann commented:
Thanks @caleb15 for the Ansible documentation fix!
